File: notification-domain/src/main/java/com/example/notification/domain/shared/Locale.java
package com.example.notification.domain.shared;

import java.util.Objects;

/**
 * 사용자 언어 / 지역. 템플릿 본문이 locale 별로 다르므로 템플릿 조회 키의 일부.
 *
 * <p>BCP-47 lowercase 형식 (예: `ko-kr`, `en-us`). 시스템 기본 fallback 은 `ko-kr`.
 */
public final class Locale {

    public static final Locale KO_KR = new Locale("ko-kr");
    public static final Locale EN_US = new Locale("en-us");

    private final String tag;

    public Locale(String tag) {
        Objects.requireNonNull(tag, "locale tag must not be null");
        String normalized = tag.trim().toLowerCase();
        if (!normalized.matches("^[a-z]{2}(-[a-z0-9]{2,8})?$")) {
            throw new IllegalArgumentException("invalid locale tag: " + tag);
        }
        this.tag = normalized;
    }

    public String tag() {
        return tag;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Locale other)) return false;
        return tag.equals(other.tag);
    }

    @Override
    public int hashCode() {
        return tag.hashCode();
    }

    @Override
    public String toString() {
        return tag;
    }
}

File: notification-domain/src/main/kotlin/com/example/notification/domain/shared/Locale.kt
package com.example.notification.domain.shared

/**
 * 사용자 언어 / 지역. 템플릿 본문이 locale 별로 다르므로 템플릿 조회 키의 일부.
 *
 * BCP-47 lowercase 형식 (예: `ko-kr`, `en-us`). 시스템 기본 fallback 은 `ko-kr`.
 *
 * 생성자에서 정규화 (trim + lowercase) 한 값을 보존하므로 equals/hashCode 는 직접 구현.
 */
class Locale(tag: String) {

    @get:JvmName("tag")
    val tag: String = tag.trim().lowercase()

    init {
        require(this.tag.matches(TAG_PATTERN)) { "invalid locale tag: $tag" }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Locale) return false
        return tag == other.tag
    }

    override fun hashCode(): Int = tag.hashCode()

    override fun toString(): String = tag

    companion object {
        // KO_KR / EN_US 보다 먼저 초기화되어야 함 — 아래 상수 생성자가 이 패턴을 참조.
        private val TAG_PATTERN = Regex("^[a-z]{2}(-[a-z0-9]{2,8})?$")

        @JvmField
        val KO_KR: Locale = Locale("ko-kr")

        @JvmField
        val EN_US: Locale = Locale("en-us")
    }
}
